File: remote/fsharp-ws-server/PipelineStages.fs
module OresSoftware.Dd.FsWs.PipelineStages

open System
open System.Text.Json
open System.Text.Json.Nodes
open System.Threading

/// The actual stage logic, shared verbatim between the Rx.NET and native-async pipeline
/// implementations. Each stage is a plain F# function that either returns a value or raises.
///
/// The point of the comparison this module exists for is that *only* the orchestration
/// differs between the two implementations — the per-stage work is identical, byte for
/// byte, so any performance / debuggability difference observed downstream is
/// attributable to the coordination library, not to the work itself.
///
/// The five stages model a realistic-ish WebSocket request pipeline:
///
///   1. `parse`     — decode an incoming text frame as JSON.
///   2. `validate`  — schema-check; reject if `id` or `payload` are missing.
///   3. `lookupA` / `lookupB` — two fan-out simulated downstream lookups (each sleeps a
///      few ms to mimic an HTTP / DB hop).
///   4. `score`     — combine the parent record with both enrichments into a score.
///   5. `serialize` — encode back to JSON text.
///
/// `poison` is included so the stack-trace comparison reliably has a failure to exercise.

/// Simulated downstream-lookup latency, ms. Kept short so unit tests stay fast.
[<Literal>]
let LOOKUP_LATENCY_MIN_MS = 1

[<Literal>]
let LOOKUP_LATENCY_MAX_MS = 4

let private randomLatencyMs () =
    // Inclusive upper bound: [min, max]
    Random.Shared.Next(LOOKUP_LATENCY_MIN_MS, LOOKUP_LATENCY_MAX_MS + 1)

let private sleepMs (ms: int) =
    if ms > 0 then Thread.Sleep ms

let parse (input: string) : JsonNode =
    if isNull input then
        raise (ArgumentNullException(nameof input))
    match JsonNode.Parse(input) with
    | null -> raise (JsonException "parse: empty JSON input")
    | node -> node

let validate (parsed: JsonNode) : JsonNode =
    if isNull parsed then
        raise (ArgumentException "validate: parsed payload is null")
    let obj = parsed.AsObject()
    if not (obj.ContainsKey "id") then
        raise (ArgumentException "validate: missing required field `id`")
    if not (obj.ContainsKey "payload") then
        raise (ArgumentException "validate: missing required field `payload`")
    parsed

let enrichLookupA (validated: JsonNode) : string =
    sleepMs (randomLatencyMs ())
    let id = validated.["id"].GetValue<string>()
    // hashCode-equivalent — System.String.GetHashCode is randomised across runs in
    // .NET, so we roll our own deterministic 16-bit hash so the output is comparable
    // across the Java/F# implementations.
    let mutable h = 0
    for ch in id do h <- h * 31 + int ch
    sprintf "lookupA[%s]=%d" id (h &&& 0xffff)

let enrichLookupB (validated: JsonNode) : string =
    sleepMs (randomLatencyMs ())
    let id = validated.["id"].GetValue<string>()
    let payloadLen = validated.["payload"].ToJsonString().Length
    sprintf "lookupB[%s]=%d" id payloadLen

/// Deliberate exception thrown from inside `score` for the stack-trace comparison.
/// Kept as a separate function so the trace shows `PipelineStages.poison` at the
/// bottom, demonstrating that whatever the coordination library does on top, the
/// *cause* of the failure is always discoverable in the trace.
let poison (validated: JsonNode) : exn =
    let id = validated.["id"].GetValue<string>()
    InvalidOperationException(sprintf "score: deliberate poison-pill id=%s" id) :> exn

let score (validated: JsonNode) (lookupA: string) (lookupB: string) : JsonNode =
    let id = validated.["id"].GetValue<string>()
    if id = "poison" then
        raise (poison validated)
    let composite = lookupA.Length * 31 + lookupB.Length
    let out = JsonObject()
    out.["id"]      <- JsonValue.Create(id)
    out.["score"]   <- JsonValue.Create(composite)
    out.["lookupA"] <- JsonValue.Create(lookupA)
    out.["lookupB"] <- JsonValue.Create(lookupB)
    out :> JsonNode

let serialize (scored: JsonNode) : string =
    scored.ToJsonString()

File: remote/deployments/akka-ws-server/src/main/java/com/oresoftware/dd/akkaws/pipeline/PipelineStages.java
package com.oresoftware.dd.akkaws.pipeline;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.Objects;
import java.util.concurrent.ThreadLocalRandom;

/**
 * The actual stage logic, shared verbatim between the async.java and Akka Streams pipeline
 * implementations. Each stage is a plain Java function that either returns a value or throws.
 *
 * <p>The point of the comparison this module exists for is that <em>only</em> the orchestration
 * differs between the two implementations — the per-stage work is identical, byte for byte, so
 * any performance / debuggability difference observed downstream is attributable to the
 * coordination library, not to the work itself.
 *
 * <p>The five stages model a realistic-ish WebSocket request pipeline:
 *
 * <ol>
 *   <li>{@link #parse(String)} — decode an incoming text frame as JSON.</li>
 *   <li>{@link #validate(JsonNode)} — schema-check; reject if {@code id} or {@code payload}
 *       are missing.</li>
 *   <li>{@link #enrichLookupA(JsonNode)} / {@link #enrichLookupB(JsonNode)} — two fan-out
 *       simulated downstream lookups (each sleeps a few ms to mimic an HTTP / DB hop).</li>
 *   <li>{@link #score(JsonNode, String, String)} — combine the parent record with both
 *       enrichments into a score.</li>
 *   <li>{@link #serialize(JsonNode)} — encode back to JSON text.</li>
 * </ol>
 *
 * <p>{@link #poison(JsonNode)} is included so the
 * {@code StackTraceComparisonTest} can deliberately throw from inside the score stage and the
 * two pipelines' stack-trace shapes can be diffed.
 */
public final class PipelineStages {

  private PipelineStages() {
  }

  private static final ObjectMapper MAPPER = new ObjectMapper();

  /** Simulated downstream-lookup latency, ms. Kept short so unit tests stay fast. */
  static final int LOOKUP_LATENCY_MIN_MS = 1;
  static final int LOOKUP_LATENCY_MAX_MS = 4;

  public static JsonNode parse(final String input) throws Exception {
    Objects.requireNonNull(input, "input");
    return MAPPER.readTree(input);
  }

  public static JsonNode validate(final JsonNode parsed) {
    if (parsed == null || !parsed.has("id")) {
      throw new IllegalArgumentException("validate: missing required field `id`");
    }
    if (!parsed.has("payload")) {
      throw new IllegalArgumentException("validate: missing required field `payload`");
    }
    return parsed;
  }

  public static String enrichLookupA(final JsonNode validated) {
    sleepMs(randomLatencyMs());
    return "lookupA[" + validated.get("id").asText() + "]=" + (validated.get("id").asText().hashCode() & 0xffff);
  }

  public static String enrichLookupB(final JsonNode validated) {
    sleepMs(randomLatencyMs());
    return "lookupB[" + validated.get("id").asText() + "]=" + (validated.get("payload").toString().length());
  }

  public static JsonNode score(final JsonNode validated, final String lookupA, final String lookupB) {
    if ("poison".equals(validated.path("id").asText())) {
      // Deliberate trigger for stack-trace comparison test. Mimics a real bug deep in the
      // last business stage.
      throw poison(validated);
    }
    final int composite = lookupA.length() * 31 + lookupB.length();
    final ObjectNode out = MAPPER.createObjectNode();
    out.put("id", validated.get("id").asText());
    out.put("score", composite);
    out.put("lookupA", lookupA);
    out.put("lookupB", lookupB);
    return out;
  }

  public static String serialize(final JsonNode scored) throws Exception {
    return MAPPER.writeValueAsString(scored);
  }

  // --- helpers ---------------------------------------------------------------------------

  /**
   * Deliberate exception thrown from inside {@link #score} for the
   * {@code StackTraceComparisonTest}. Kept as a separate method so the stack trace shows
   * `PipelineStages.poison(...)` at the bottom, demonstrating that whatever the coordination
   * library does on top, the *cause* of the failure is always discoverable in the trace.
   */
  public static RuntimeException poison(final JsonNode validated) {
    return new IllegalStateException("score: deliberate poison-pill id=" + validated.path("id").asText());
  }

  private static int randomLatencyMs() {
    return ThreadLocalRandom.current().nextInt(LOOKUP_LATENCY_MIN_MS, LOOKUP_LATENCY_MAX_MS + 1);
  }

  private static void sleepMs(final long ms) {
    if (ms <= 0) {
      return;
    }
    try {
      Thread.sleep(ms);
    } catch (InterruptedException ie) {
      Thread.currentThread().interrupt();
    }
  }
}
